src/librustc/ty/sty.rs

```diff
@@ -1288,6 +1288,27 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
+    pub fn conservative_is_uninhabited(&self) -> bool {
+        // Uncontentious uninhabitableness check
+        match self.sty {
+            ty::TyNever => true,
+            // We could extend this to ADTs in which every variant is uninhabited
+            // (for enums, unions, etc.).
+            ty::TyAdt(def, _) => def.variants.is_empty(),
+            ty::TyTuple(tys) => tys.iter().any(|ty| ty.conservative_is_uninhabited()),
+            ty::TyArray(ty, len) => {
```

Comment:
Any particular reason to expand the set of cases here? Not that I I necessarily object. It seems though that including `(!,)` but not `struct Foo(!)` is sort of surprising to me -- in other words, I'm tempted to stick to `!` and `Void`, and separate out this aspect of the PR.